### PR TITLE
Add mode bit to track whether we are in safe nulls mode

### PR DESCRIPTION
**Pull request by odersky**
The bit is set if

 - -Yexplicit-nulls is set
 - we are not in an `import.unsafeNulls` scope

### CODE REVIEW

**Comment by odersky** [comment]
@noti0na1 This might be a good basis for the explicit-nulls PR.

**Comment by noti0na1** [comment]
@odersky Thanks! I will take a look

**Comment by noti0na1** [comment]
I have merged this branch into my PR. 
One small modification is required: when we add the SafeNulls bit, we also need to check the unsafeNulls language feature is not set globally.
It seems we can use the new Mode bit to check UnsafeNulls at all places. The StickyKey is no longer needed as well.

File: compiler/src/dotty/tools/dotc/Run.scala [hunk 1]
```diff
@@ -71,11 +71,13 @@ class Run(comp: Compiler, ictx: Context) extends ImplicitRunInfo with Constraint
       .setPeriod(Period(comp.nextRunId, FirstPhaseId))
       .setScope(rootScope)
     rootScope.enter(ctx.definitions.RootPackage)(using bootstrap)
-    val start = bootstrap.fresh
+    var start = bootstrap.fresh
       .setOwner(defn.RootClass)
       .setTyper(new Typer)
       .addMode(Mode.ImplicitsEnabled)
       .setTyperState(ctx.typerState.fresh(ctx.reporter))
+    if ctx.settings.YexplicitNulls.value then
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/Run.scala:79]: Also need `&& !Feature.enabledBySetting(nme.unsafeNulls)`

Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/Run.scala:79]: Also need `&& !Feature.enabledBySetting(nme.unsafeNulls)`

File: compiler/src/dotty/tools/dotc/core/Contexts.scala [hunk 2]
```diff
@@ -632,7 +628,14 @@ object Contexts {
     def setRun(run: Run): this.type = updateStore(runLoc, run)
     def setProfiler(profiler: Profiler): this.type = updateStore(profilerLoc, profiler)
     def setNotNullInfos(notNullInfos: List[NotNullInfo]): this.type = updateStore(notNullInfosLoc, notNullInfos)
-    def setImportInfo(importInfo: ImportInfo): this.type = updateStore(importInfoLoc, importInfo)
+    def setImportInfo(importInfo: ImportInfo): this.type =
+      importInfo.mentionsFeature(nme.unsafeNulls) match
+        case Some(true) =>
+          setMode(this.mode &~ Mode.SafeNulls)
+        case Some(false) if ctx.settings.YexplicitNulls.value =>
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/core/Contexts.scala:635]: Also need `&& !Feature.enabledBySetting(nme.unsafeNulls)`

**Review by noti0na1 — CHANGES_REQUESTED** [review]

File: compiler/src/dotty/tools/dotc/core/Contexts.scala (hunk 2 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/Contexts.scala:635]: I think not. The initial setting of `unsafeNulls` is irrelevant here.

Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/Contexts.scala:635]: If we don't have the test and `unsafeNulls` is enabled globally, the `SafeNulls` will be added to the context here, which is not expected.

Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/Contexts.scala:635]: No, I think that _is_ the expected behavior. A local (un-)import overrides an outer import and a global setting.

Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/Contexts.scala:635]: I think you are right. We should allow local imports to overrides a global setting.

**Review by noti0na1 — APPROVED** [review]

**Pull request merged**